ducktape/services/service.py

```diff
@@ -52,21 +60,71 @@ def start(self):
                     "Service: %s, node.account: %s" % (self.__class__.__name__, str(node.account)))
             node.account.set_logger(self.logger)
 
-            self.logger.debug("Forcibly cleaning node %d on %s", idx, node.account.hostname)
-            self._clean_node(node)
+    def start(self):
+        """Start the service on all nodes."""
+        self.logger.info("Starting service: %s" % self.__class__.__name__)
+
+        self.logger.debug("Killing processes and attempting to clean up before starting nodes.")
+        for node in self.nodes:
+            # Added precaution - kill running processes
+            self._kill_running_processes(node)
+
+            self.stop_node(node)
+            self.clean_node(node)
+
+        self.logger.info("Starting nodes for %s " % self.__class__.__name__)
+        for node in self.nodes:
+            self.logger.debug("Starting node %d on %s" % (self.idx(node), node.account.hostname))
+            self.start_node(node)
+            self.wait_until_alive(node)
+
+    def start_node(self, node):
+        """Start service process(es) on the given node."""
+        raise NotImplementedError("Subclasses must implement clean_node.")
+
+    def wait_until_alive(self, node):
+        """Wait until this service node is alive.
+        """
+        pass
 
     def wait(self):
-        """Wait for the service to finish. This only makes sense for tasks with a fixed
-        amount of work to do. For services that generate output, it is only
-        guaranteed to be available after this call returns.
+        """Wait for the service to finish.
+        This only makes sense for tasks with a fixed amount of work to do. For services that generate
+        output, it is only guaranteed to be available after this call returns.
         """
         pass
 
     def stop(self):
-        """If the service left any running processes or data, clean them up."""
-        for idx, node in enumerate(self.nodes, 1):
+        """Stop service processes on each node in this service.
+        Subclasses must override stop_node.
+        """
+        for node in self.nodes:
+            self.stop_node(node)
             node.account.set_logger(None)
 
+    def stop_node(self, node):
+        """Halt service process(es) on this node."""
+        raise NotImplementedError("Subclasses must implement stop_node.")
+
+    def clean(self):
+        """Clean up persistent state on each node - e.g. logs, config files etc.
+        Subclasses must override clean_node.
+        """
+        for node in self.nodes:
+            self.clean_node(node)
+
+    def clean_node(self, node):
+        """Clean up persistent state on this node - e.g. service logs, configuration files etc."""
+        raise NotImplementedError("Subclasses must implement clean_node.")
```

Comment:
Question here is whether we enforce good practice by raising this as the default or make writing tests easier by allowing services that can handle old state without a problem (e.g. they just overwrite old logs) ignore writing this method. Probably not a big deal since they can always override and leave it empty, but something to think about.